internal/hcsoci/resources.go

```diff
@@ -23,158 +20,81 @@ func (r *Resources) NetNS() string {
-		for len(r.vsmbMounts) != 0 {
-			mount := r.vsmbMounts[len(r.vsmbMounts)-1]
-			if err := vm.RemoveVSMB(ctx, mount); err != nil {
-				return err
-			}
-			r.vsmbMounts = r.vsmbMounts[:len(r.vsmbMounts)-1]
-		}
-
-		for len(r.pipeMounts) != 0 {
-			mount := r.pipeMounts[len(r.pipeMounts)-1]
-			if err := vm.RemovePipe(ctx, mount); err != nil {
-				return err
+	if vm != nil {
+		if r.addedNetNSToVM {
+			if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
+				log.G(ctx).Warn(err)
 			}
-			r.pipeMounts = r.pipeMounts[:len(r.pipeMounts)-1]
+			r.addedNetNSToVM = false
 		}
 
-		for len(r.plan9Mounts) != 0 {
-			mount := r.plan9Mounts[len(r.plan9Mounts)-1]
-			if err := vm.RemovePlan9(ctx, mount); err != nil {
-				return err
+		// vpci cleanup
+		for _, name := range r.vpciDevices {
+			if err := vm.RemoveDevice(ctx, name); err != nil {
+				log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 			}
-			r.plan9Mounts = r.plan9Mounts[:len(r.plan9Mounts)-1]
 		}
 
-		for _, sm := range r.scsiMounts {
-			if err := vm.RemoveSCSI(ctx, sm.path); err != nil {
-				return err
-			}
-			if sm.autoManage {
-				if err := os.Remove(sm.path); err != nil {
-					log.G(ctx).WithError(err).Warnf("failed to remove automanage-virtual-disk at: %q", sm.path)
+		if all {
+			// For now the resources that implement the ReleaseCloser interface are vsmb shares,
+			// scsi mounts, pipe mounts, and plan9 mounts. The goal is to move over networking,
+			// vpci devices, and layers to this so they all follow the same format.
+			for len(r.resources) != 0 {
+				resource := r.resources[len(r.resources)-1]
+				if resource != nil {
+					if err := resource.Release(ctx); err != nil {
+						return err
+					}
 				}
+				r.resources = r.resources[:len(r.resources)-1]
 			}
 		}
-		r.scsiMounts = nil
-	}
 
-	if vm != nil && vm.DeleteContainerStateSupported() {
-		if err := vm.DeleteContainerState(ctx, r.id); err != nil {
-			log.G(ctx).WithError(err).Error("failed to delete container state")
+		// cleanup container state
+		if vm.DeleteContainerStateSupported() {
+			if err := vm.DeleteContainerState(ctx, r.id); err != nil {
+				log.G(ctx).WithError(err).Error("failed to delete container state")
+			}
 		}
 	}
 
-	if len(r.layers) != 0 {
-		op := UnmountOperationSCSI
-		if vm == nil || all {
-			op = UnmountOperationAll
-		}
-		var crp string
-		if vm != nil {
-			crp = containerRootfsPath(vm, r.containerRootInUVM)
-		}
-		err := UnmountContainerLayers(ctx, r.layers, crp, vm, op)
-		if err != nil {
+	// network endpoint cleanup
+	if r.createdNetNS {
+		if err := uvm.CloseAllNetworkEndpoints(ctx, r.netNS, r.networkEndpoints); err != nil {
 			return err
 		}
-		r.layers = nil
+		r.createdNetNS = false
 	}
 
-	if vm != nil {
-		for _, name := range r.vpciDevices {
-			if err := vm.RemoveDevice(ctx, name); err != nil {
-				log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
+	// layers cleanup
+	if r.layers != nil {
+		if len(r.layers.Layers) != 0 {
```

Comment:
Do we need this double check here? Can we not just check that r.layers is not nil? This extra case will not really do anything since the Close call will still return an error on UnmountContainerLayers when there are less than two layers already.